piptools/utils.py

```diff
@@ -522,3 +531,129 @@ def parse_requirements_from_wheel_metadata(
             markers=parts.markers,
             extras=parts.extras,
         )
+
+
+def callback_config_file_defaults(
+    ctx: click.Context, param: click.Parameter, value: str | None
+) -> Path | None:
+    """
+    Returns the path to the config file with defaults being used, or `None` if no such file is
+    found.
+
+    Defaults for `click.Command` parameters should be override-able in a config file. `pip-tools`
+    will use the first file found, searching in this order: an explicitly given config file, a
+    `.pip-tools.toml`, a `pyproject.toml` file. Those files are searched for in the same directory
+    as the requirements input file.
+    """
+    if value is None:
+        config_file = select_config_file(ctx.params.get("src_files", ()))
+        if config_file is None:
+            return None
+    else:
+        config_file = Path(value)
+
+    if config := parse_config_file(config_file):
+        _assign_config_to_cli_context(ctx, config)
+    else:
+        return None
```

Comment:
@j00bar through a series of code conversions I discovered that the check here is rather weird — `if (not) config:` checks for a falsy value (which a present but empty section would be).
This seems to be semantically incorrect. If there's an empty config, we'd still want to recognize its existence.
Is this the reason you thought it'd be hard to decouple this?
Why do we even want to pretend a config file wasn't passed at all to the CLI when we don't see a config there? I don't think we should do that..

Reply:
You're correct - thank you.

Reply:
@j00bar looks like this behavior hasn't been corrected.